test/modules/layers/test_quantization.py

```diff
@@ -84,23 +87,72 @@ def test_preprocess(self):
         actual_flat_shape = torch.tensor(encoded_flat.shape)
         actual_permuted_shape = torch.tensor(permuted_shape)
 
-        assert torch.equal(
-            actual_flat_shape, expected_flat_shape
-        ), f"actual flattened shape: {actual_flat_shape}, expected flattened shape: {expected_flat_shape}"
+        assert_expected(actual_flat_shape, expected_flat_shape)
 
-        assert torch.equal(
-            actual_permuted_shape, expected_permuted_shape
-        ), f"actual permuted shape: {actual_permuted_shape}, expected permuted shape: {expected_permuted_shape}"
+        assert_expected(actual_permuted_shape, expected_permuted_shape)
 
     def test_preprocess_channel_dim_assertion(self):
         with self.assertRaises(ValueError):
-            encoded_flat, permuted_shape = self.vq._preprocess(self.encoded[:, :4, :])
+            _, _ = self.vq._preprocess(self.encoded[:, :4, :])
 
     def test_postprocess(self):
         quantized = self.vq._postprocess(self.input_tensor_flat, torch.Size([2, 2, 3]))
         actual_quantized_shape = torch.tensor(quantized.shape)
         expected_quantized_shape = torch.tensor([2, 3, 2])
 
-        assert torch.equal(
-            actual_quantized_shape, expected_quantized_shape
-        ), f"actual quantized shape: {actual_quantized_shape}, expected quantized shape: {expected_quantized_shape}"
+        assert_expected(actual_quantized_shape, expected_quantized_shape)
+
+    def test_init_embedding_and_preprocess(self):
+        assert not self.vq._is_embedding_init, "embedding init flag not False initially"
+
+        _, _ = self.vq._init_embedding_and_preprocess(self.encoded)
+
+        assert self.vq._is_embedding_init, "embedding init flag not True after init"
+
+        actual_weight = self.vq.embedding
+        expected_weight = torch.Tensor(
+            [
+                [2.0, -1.0, 0.0, 2.0, 0.0],
+                [2.0, 1.0, 0.0, 1.0, 1.0],
+                [0.0, 1.0, -1.0, 2.0, -1.0],
+                [1.0, 0.0, -1.0, -1.0, 1.0],
+            ]
+        )
+        assert_expected(actual_weight, expected_weight)
+
+        actual_code_avg = self.vq.code_avg
+        expected_code_avg = actual_weight
+        assert_expected(actual_code_avg, expected_code_avg)
+
+        actual_code_usage = self.vq.code_usage
+        expected_code_usage = torch.ones(self.num_embeddings)
+        assert_expected(actual_code_usage, expected_code_usage)
+
+    def test_ema_update(self):
```

Comment:
It might also make sense to add a unit test where you call .backward() to confirm that the buffers behave as expected

Reply:
good call

Reply:
Agreed